src/worker/mod.rs

```diff
@@ -437,15 +431,22 @@ impl Worker {
     /// Returns `Ok(None)` in the event of a poll timeout, or if there was some gRPC error that
     /// callers can't do anything about.
     async fn workflow_poll(&self) -> Result<Option<ValidPollWFTQResponse>, PollWfError> {
+        // We can't say we're shut down if there are outstanding LAs, as they could end up WFT
+        // heartbeating which is a "new" workflow task that we need to accept and process as long
+        // as the LA is outstanding.
+        if *self.wfts_drained.borrow() && self.local_act_mgr.num_outstanding() == 0 {
+            return Err(PollWfError::ShutDown);
+        }
+
         let sem = self
             .workflows_semaphore
             .acquire()
             .await
             .expect("outstanding workflow tasks semaphore not dropped");
 
         let res = self
-            .wf_task_poll_buffer
-            .poll()
+            .wf_task_source
+            .next_wft()
```

Comment:
This is the reason for switching drained on and off. The idea is that it won't return an error. It shouldn't if you still have running LAs. If you don't, we won't ask for WFTs from completion. I'll try to figure out making drained permanent and this and will add a specific test, which is one of the things I needed to follow up on anyway

Reply:
OK So this case is covered by `local_act_heartbeat::case_1` -- the general case of having *any* wft complete which returned a new task is technically not, since the only case where that happens is during WFT heartbeating which the test covers, so I just don't have a great way to set up such a test, but it should be handled by the changes to the condition for when `ShutDown` is returned by `workflow_poll`.

If things ever change such that returning a WFT during a complete is enabled in more scenarios, then tests will be added to test that path, and with shutdown as well